File: crates/minml-core/src/cpu/backend.rs
use crate::buffer::Buffer;
use crate::device::Device;
use crate::error::{MinmlError, Result};
use parking_lot::RwLock;
use std::any::Any;
use std::sync::Arc;

// Plain heap-allocated bytes. RwLock (not Mutex) so multiple kernel reads
// over the same buffer don't deadlock — important when the same Array is
// passed twice to an op (e.g. dot(xy, xy)). Writes (h2d, output buffers)
// are exclusive. parking_lot::RwLock supports multiple readers on the
// same thread, which is exactly what we need.
pub struct CpuBuffer {
    pub(crate) data: RwLock<Vec<u8>>,
    bytes: usize,
}

impl CpuBuffer {
    pub fn new(bytes: usize) -> Self {
        Self {
            data: RwLock::new(vec![0u8; bytes]),
            bytes,
        }
    }
}

impl Buffer for CpuBuffer {
    fn bytes(&self) -> usize {
        self.bytes
    }
    fn device(&self) -> Device {
        Device::Cpu
    }
    fn as_any(&self) -> &dyn Any {
        self
    }
}

pub fn allocate(bytes: usize) -> Arc<dyn Buffer> {
    Arc::new(CpuBuffer::new(bytes))
}

fn as_cpu(b: &dyn Buffer) -> &CpuBuffer {
    b.as_any().downcast_ref::<CpuBuffer>().expect("cpu buffer")
}

pub fn copy_host_to_buffer(dst: &dyn Buffer, src: &[u8]) -> Result<()> {
    let cpu = as_cpu(dst);
    let mut data = cpu.data.write();
    if src.len() > data.len() {
        return Err(MinmlError::Other(format!(
            "h2d size mismatch: src={} dst={}",
            src.len(),
            data.len()
        )));
    }
    data[..src.len()].copy_from_slice(src);
    Ok(())
}

pub fn copy_buffer_to_host(src: &dyn Buffer, dst: &mut [u8]) -> Result<()> {
    let cpu = as_cpu(src);
    let data = cpu.data.read();
    if dst.len() > data.len() {
        return Err(MinmlError::Other(format!(
            "d2h size mismatch: src={} dst={}",
            data.len(),
            dst.len()
        )));
    }
    dst.copy_from_slice(&data[..dst.len()]);
    Ok(())
}

// Internal helpers: typed views over the buffer bytes for kernels.
pub(crate) fn with_f32_mut<F, R>(buf: &dyn Buffer, f: F) -> R
where
    F: FnOnce(&mut [f32]) -> R,
{
    let cpu = as_cpu(buf);
    let mut bytes = cpu.data.write();
    f(bytemuck::cast_slice_mut::<u8, f32>(&mut bytes))
}

pub(crate) fn with_f32<F, R>(buf: &dyn Buffer, f: F) -> R
where
    F: FnOnce(&[f32]) -> R,
{
    let cpu = as_cpu(buf);
    let bytes = cpu.data.read();
    f(bytemuck::cast_slice::<u8, f32>(&bytes))
}

pub(crate) fn with_i32_mut<F, R>(buf: &dyn Buffer, f: F) -> R
where
    F: FnOnce(&mut [i32]) -> R,
{
    let cpu = as_cpu(buf);
    let mut bytes = cpu.data.write();
    f(bytemuck::cast_slice_mut::<u8, i32>(&mut bytes))
}

pub(crate) fn with_i32<F, R>(buf: &dyn Buffer, f: F) -> R
where
    F: FnOnce(&[i32]) -> R,
{
    let cpu = as_cpu(buf);
    let bytes = cpu.data.read();
    f(bytemuck::cast_slice::<u8, i32>(&bytes))
}

File: src/cpu/backend.ts
import type { Buffer } from "../buffer.js";
import { Device } from "../device.js";
import { MinmlError } from "../error.js";

// Plain heap-allocated bytes. JS is single-threaded so no locking is
// needed (Rust used parking_lot::RwLock to allow multiple read views of
// the same buffer when the same Array is passed twice to an op like
// dot(xy, xy)).
export class CpuBuffer implements Buffer {
  readonly data: ArrayBuffer;
  private readonly _bytes: number;

  constructor(bytes: number) {
    this._bytes = bytes;
    this.data = new ArrayBuffer(bytes);
  }

  bytes(): number {
    return this._bytes;
  }
  device(): Device {
    return Device.Cpu;
  }
}

export function allocate(bytes: number): Buffer {
  return new CpuBuffer(bytes);
}

function as_cpu(b: Buffer): CpuBuffer {
  if (!(b instanceof CpuBuffer)) {
    throw MinmlError.other("expected CpuBuffer");
  }
  return b;
}

export function copy_host_to_buffer(dst: Buffer, src: Uint8Array): void {
  const cpu = as_cpu(dst);
  if (src.length > cpu.bytes()) {
    throw MinmlError.other(
      `h2d size mismatch: src=${src.length} dst=${cpu.bytes()}`,
    );
  }
  new Uint8Array(cpu.data).set(src);
}

export function copy_buffer_to_host(src: Buffer, dst: Uint8Array): void {
  const cpu = as_cpu(src);
  if (dst.length > cpu.bytes()) {
    throw MinmlError.other(
      `d2h size mismatch: src=${cpu.bytes()} dst=${dst.length}`,
    );
  }
  dst.set(new Uint8Array(cpu.data, 0, dst.length));
}

// Typed-array views over the underlying ArrayBuffer. Multiple views over
// the same buffer share memory — same semantics as Rust's bytemuck::cast_slice.
export function f32_view(buf: Buffer): Float32Array {
  return new Float32Array(as_cpu(buf).data);
}

export function i32_view(buf: Buffer): Int32Array {
  return new Int32Array(as_cpu(buf).data);
}
